core/modules/mlfqueue.cc

```diff
@@ -0,0 +1,346 @@
+    {"set_batch_Size", "MlfqBatchArg", MODULE_CMD_FUNC(&MLFQueue::CommandBatchSize), 0},
+    {"set_max_flow_queue_size", "MlfqMaxFlowQueueSizeArg",
+          MODULE_CMD_FUNC(&MLFQueue::CommandMaxFlowQueueSize), 0}};
+
+pb_error_t MLFQueue::Init(const bess::pb::MlfqArg &arg) {
+  pb_error_t err;
+
+  if (arg.num_lvls() != 0) {
+    err = SetNumPriorityLevels(arg.num_lvls());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetNumPriorityLevels(DEFAULT_NUM_LVLS);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if (arg.max_flow_queue_size() != 0) {
+    err = SetMaxFlowQueueSize(arg.max_flow_queue_size());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetMaxFlowQueueSize(FLOW_QUEUE_MAX);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if (arg.batch_size() != 0) {
+    err = SetBatchSize(arg.batch_size());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetBatchSize(DEFAULT_BATCH_SIZE);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if(arg.init_load() != 0) {
+    load_avg_ = arg.init_load();
+  } else {
+    load_avg_ = INITIAL_LOAD;
+  }
+  ready_flows_ = 0;
+  return pb_errno(0);
+}
+
+void MLFQueue::DeInit() {
+  bess::Packet *pkt;
+
+  for (std::map<Q_Id, Flow>::iterator it=flows_.begin(); it!=flows_.end(); ) {
+    Flow *f = &it->second;
+    if (f->queue) {
+      while (llring_sc_dequeue(f->queue, (void **)&pkt) == 0) {
+        bess::Packet::Free(pkt);
+      }
+      mem_free(f->queue);
+    }
+  }
+}
+
+pb_cmd_response_t MLFQueue::CommandNumPriorityLevels(const bess::pb::MlfqLvlArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetNumPriorityLevels(arg.num_lvls()));
+  return response;
+}
+
+pb_cmd_response_t MLFQueue::CommandBatchSize(const bess::pb::MlfqBatchArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetBatchSize(arg.batch_size()));
+  return response;
+}
+
+pb_cmd_response_t MLFQueue::CommandMaxFlowQueueSize(const bess::pb::MlfqMaxFlowQueueSizeArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetMaxFlowQueueSize(arg.max_queue_size()));
+  return response;
+}
+
+void MLFQueue::ProcessBatch(bess::PacketBatch *batch) {
+  google::SetLogDestination(google::GLOG_INFO,"/vagrant/batch.log");
+  LOG(INFO)<< std::to_string(batch->cnt());
+  //insert packets in the batch into their corresponding flows
+  for (int i = 0; i < batch->cnt(); i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+    Q_Id id = MLFQueue::GetId(pkt);//doesn't handle fragmented packets
+    std::map<Q_Id, Flow>::iterator it = flows_.find(id);
+
+    //if the Flow doesn't exist create one
+    //and add the packet to the new Flow
+    if (it == flows_.end()) {
+      Flow f = {(float) max_lvl_, 0, 0, 0};
+      int slots = RoundToPowerTwo(FLOW_QUEUE_FACTOR*batch_size_);
+      int err = MLFQueue::AddQueue(&f, slots);
+      assert(err == 0);
```

Comment:
Is this assertion a substitute for an error check rather than checking an invariant?  If so, you should add a TODO saying that you need to do proper error handling here.